app.py

```diff
@@ -72,21 +27,44 @@ def get_authorization_header(root, discharge):
     """
 
     bound = Macaroon.deserialize(root).prepare_for_request(
-        Macaroon.deserialize(discharge))
+        Macaroon.deserialize(discharge)
+    )
 
     return 'Macaroon root={}, discharge={}'.format(root, bound.serialize())
 
 
+def is_connected():
+    return (
+        'openid' in flask.session or
+        'macaroon_discharge' in flask.session or
+        'macaroon_root' in flask.session
+    )
+
+
+def redirect_to_login():
+    return flask.redirect(
+        'login?next=' +
+        flask.request.url_rule.rule
+    )
+
+
 @app.route('/')
 def homepage():
-    return flask.render_template('index.html')
+    context = {}
+    if is_connected():
+        context['connected'] = True
+
+    return flask.render_template('index.html', **context)
 
 
 @app.route('/login', methods=['GET', 'POST'])
 @oid.loginhandler
 def login():
+    if is_connected():
+        return flask.redirect(oid.get_next_url())
+
     response = requests.request(
-        url='https://dashboard.snapcraft.io/dev/api/acl/',
+        url=SCA_BASE_URL + '/acl/',
```

Comment:
I guess since this is a "refactor" PR I should now be commenting on semantic choices?

I don't really like keeping this URL in a constant in this case. The use of constants in Surl led to actual confusion, as it was very hard to keep in your head which call was made to which domain. I doubt the domains are going to change often enough to warrant keeping them at the top of the file, and since the constant is only used once here at the moment, I think it would be much better to just leave the whole URL. (Also "SCA_BASE" means nothing to me. What's an SCA?)

The same applies to `UBUNTU_SSO_URL`. Although that's a better name, it's still only used once, so I think it would be much easier to read the code if the URL was just used directly.

Reply:
for the login I can understand it is never going to be used anywhere else.
But for the SCA ([software-center-agent](https://launchpad.net/software-center-agent) alias [dashboard.snapcraft.io](dashboard.snapcraft.io)) we might use for other purposes, so why not keeping it?
In this case we use it once but we never know...?

Reply:
Every time you put something in a variable you obfuscate something. Humans aren't as good as computers at keeping the contents of variables in their heads. The reason for using a variable is because something might change during the running of the code. Similarly, the reason for using a constant is because something might change in the environment the code is running in. If nothing is likely to change, I think variables and constants really make code significantly harder to read and should not be overused.

In Surl, Celso had a good reason for using a variable, because he was swapping out `dashboard.staging.snapcraft.io` for `dashboard.snapcraft.io`. But in our case, I we don't need to use the staging domain. We should just keep it simple.

I also think URLs are *especially* useful to see in their entirety. Think about cases where this variable might effect you:

- You're reading the code, trying to understand how it interacts with the API
- You're writing new code, using a new endpoint on the same API

In both cases, I think your job would be made much easier by just seeing the URL directly, rather than having to keep scrolling to the top of the file to re-compose the URL in your head. As I said, I think this already caused confusion when we were reading the Surl code, 'cos it was hard to tell which calls were made to login.ubuntu.com and which to dashboard.snapcraft.io.

Reply:
ok I understand your point, I will remove those constants. Let's keep this code simple and easily readable